vm/src/value.cpp

```diff
@@ -0,0 +1,533 @@
+#include "value.h"
+#include "deserializer.h"
+#include "matrix_3.h"
+#include "matrix_4.h"
+#include "serializer.h"
+#include "vector_2.h"
+#include "vector_3.h"
+#include <cassert>
+#include <utility>
+
+namespace Borges {
+
+Value::Value(bool b): type_(Type::boolean), boolean_(b)
+{
+}
+
+Value::Value(double x) : type_(Type::number), number_(x)
+{
+}
+
+Value::Value(const Vector_2& v)
+    : type_(Type::vector_2),
+      vector_2_(std::unique_ptr<const Vector_2>(new Vector_2(v)))
+{
+}
+
+Value::Value(const Vector_3& v)
+    : type_(Type::vector_3),
+      vector_3_(std::unique_ptr<const Vector_3>(new Vector_3(v)))
+{
+}
+
+Value::Value(const Matrix_3& m)
+    : type_(Type::matrix_3),
+      matrix_3_(std::unique_ptr<const Matrix_3>(new Matrix_3(m)))
+{
+}
+
+Value::Value(const Matrix_4& m)
+    : type_(Type::matrix_4),
+      matrix_4_(std::unique_ptr<const Matrix_4>(new Matrix_4(m)))
+{
+}
+
+Value::Value(std::vector<bool> bs)
+    : type_(Type::boolean_array),
+      boolean_array_(std::shared_ptr<const std::vector<bool>>(
+          new std::vector<bool>(std::move(bs))))
+{
+}
+
+Value::Value(std::vector<double> xs)
+    : type_(Type::number_array),
+      number_array_(std::shared_ptr<const std::vector<double>>(
+          new std::vector<double>(std::move(xs))))
+{
+}
+
+Value::Value(std::vector<Vector_2> vs)
+    : type_(Type::vector_2_array),
+      vector_2_array_(std::shared_ptr<const std::vector<Vector_2>>(
+          new std::vector<Vector_2>(std::move(vs))))
+{
+}
+
+Value::Value(std::vector<Vector_3> vs)
+    : type_(Type::vector_3_array),
+      vector_3_array_(std::shared_ptr<const std::vector<Vector_3>>(
+          new std::vector<Vector_3>(std::move(vs))))
+{
+}
+
+Value::Value(std::vector<Matrix_3> ms)
+    : type_(Type::matrix_3_array),
+      matrix_3_array_(std::shared_ptr<const std::vector<Matrix_3>>(
+          new std::vector<Matrix_3>(std::move(ms))))
+{
+}
+
+Value::Value(std::vector<Matrix_4> ms)
+    : type_(Type::matrix_4_array),
+      matrix_4_array_(std::shared_ptr<const std::vector<Matrix_4>>(
+          new std::vector<Matrix_4>(std::move(ms))))
+{
+}
+
+Value::Value(const Value& x)
+{
+    construct(x);
+}
+
+Value::Value(Value&& x)
+{
+    construct(x);
+}
+
+Value&
+Value::operator=(const Value& x)
+{
+    if (this != &x) {
+        destroy();
+        construct(x);
+    }
+    return *this;
+}
+
+Value&
+Value::operator=(Value&& x)
+{
```

Comment:
doesn't a referenced reference implicitly get cast to a reference making this assignment operator redundant ?